File: cpp/src/ui/menu.cpp
#include "menu.hpp"

namespace ui {

void Menu::addEntry(MenuEntry entry) {
    m_entries.push_back(std::move(entry));
}

void Menu::addEntry(std::string name, bool selectable, std::function<void()> onSelected) {
    m_entries.push_back({std::move(name), selectable, std::move(onSelected)});
}

void Menu::addSeparator() {
    m_entries.push_back(MenuEntry::Separator());
}

void Menu::moveUp() {
    if (m_entries.empty()) return;

    if (m_at < 0) {
        for (size_t i = 0; i < m_entries.size(); i++) {
            if (m_entries[i].selectable) {
                m_at = static_cast<int>(i);
                break;
            }
        }
    } else if (m_at > 0) {
        for (int i = m_at - 1; i >= 0; i--) {
            if (m_entries[i].selectable) {
                m_at = i;
                break;
            }
        }
    }
}

void Menu::moveDown() {
    if (m_entries.empty()) return;

    int count = static_cast<int>(m_entries.size());

    if (m_at < count - 1) {
        for (int i = m_at + 1; i < count; i++) {
            if (m_entries[i].selectable) {
                m_at = i;
                break;
            }
        }
    } else {
        for (int i = count - 1; i >= 0; i--) {
            if (m_entries[i].selectable) {
                m_at = i;
                break;
            }
        }
    }
}

void Menu::select() {
    if (m_entries.empty()) return;

    if (m_at < 0) {
        m_at = 0;
    } else if (m_at >= static_cast<int>(m_entries.size())) {
        m_at = static_cast<int>(m_entries.size()) - 1;
    }

    if (m_entries[m_at].onSelected) {
        m_entries[m_at].onSelected();
    }
}

void Menu::draw(tui::Screen& screen) {
    if (m_entries.empty()) return;

    int maxLen = 0;
    for (size_t i = 0; i < m_entries.size(); i++) {
        if (!m_entries[i].name.empty()) {
            int len = static_cast<int>(m_entries[i].name.size()) + 2;
            if (len > maxLen) maxLen = len;
        }
    }

    int startX = m_x;
    int startY = m_y;
    if (startX == 0) {
        startX = (screen.width() - maxLen) / 2;
    }
    if (startY == 0) {
        startY = (screen.height() - static_cast<int>(m_entries.size())) / 2;
    }

    for (size_t i = 0; i < m_entries.size(); i++) {
        int y = startY + static_cast<int>(i);

        if (m_entries[i].name.empty()) {
            continue;
        }

        screen.putString(startX, y, m_entries[i].name);

        auto fg = m_entries[i].selectable ? tui::Color::White() : tui::Color::Gray();
        int len = static_cast<int>(m_entries[i].name.size());
        for (int j = 0; j < len; j++) {
            screen.setFgColor(startX + j, y, fg);
        }

        if (static_cast<int>(i) == m_at) {
            screen.putChar(startX + len + 1, y, "<");
            screen.setFgColor(startX + len + 1, y, tui::Color::Yellow());
        }
    }
}

} // namespace ui

File: rust/src/ui/menu.rs
//! Menu widget

use crate::tui::{cell::Color, screen::Screen};

/// Menu entry
pub struct Entry {
    pub name: Option<String>,
    pub selectable: bool,
    pub callback: Option<Box<dyn FnMut() + Send>>,
}

/// Menu widget
pub struct Menu {
    entries: Vec<Entry>,
    selected: i32,
}

impl Menu {
    pub fn new() -> Self {
        Menu {
            entries: Vec::new(),
            selected: -1,
        }
    }

    pub fn add_entry<F>(&mut self, name: &str, selectable: bool, callback: Option<F>)
    where
        F: FnMut() + Send + 'static,
    {
        self.entries.push(Entry {
            name: Some(name.to_string()),
            selectable,
            callback: callback.map(|f| Box::new(f) as Box<dyn FnMut() + Send>),
        });

        // Auto-select first selectable entry
        if selectable && self.selected < 0 {
            self.selected = (self.entries.len() - 1) as i32;
        }
    }

    pub fn add_separator(&mut self) {
        self.entries.push(Entry {
            name: None,
            selectable: false,
            callback: None,
        });
    }

    pub fn move_up(&mut self) {
        if self.selected < 0 {
            // Find first selectable
            for (i, entry) in self.entries.iter().enumerate() {
                if entry.selectable {
                    self.selected = i as i32;
                    break;
                }
            }
        } else if self.selected > 0 {
            for i in (0..self.selected as usize).rev() {
                if self.entries[i].selectable {
                    self.selected = i as i32;
                    break;
                }
            }
        }
    }

    pub fn move_down(&mut self) {
        let count = self.entries.len() as i32;
        if self.selected < count - 1 {
            for i in (self.selected as usize + 1)..self.entries.len() {
                if self.entries[i].selectable {
                    self.selected = i as i32;
                    break;
                }
            }
        }
    }

    pub fn select(&mut self) {
        if self.selected >= 0 && (self.selected as usize) < self.entries.len() {
            if let Some(ref mut callback) = self.entries[self.selected as usize].callback {
                callback();
            }
        }
    }

    pub fn draw_on(&self, screen: &mut Screen) {
        // Calculate max width
        let max_len = self
            .entries
            .iter()
            .filter_map(|e| e.name.as_ref())
            .map(|n| n.len() + 2)
            .max()
            .unwrap_or(10);

        // Calculate position (centered)
        let start_x = (screen.width().saturating_sub(max_len)) / 2;
        let start_y = (screen.height().saturating_sub(self.entries.len())) / 2;

        // Draw entries
        for (i, entry) in self.entries.iter().enumerate() {
            let y = start_y + i;

            if let Some(ref name) = entry.name {
                screen.put_str(start_x, y, name);

                // Set color
                let fg = if entry.selectable {
                    Color::WHITE
                } else {
                    Color::GRAY
                };
                for j in 0..name.len() {
                    screen.set_fg_color(start_x + j, y, fg);
                }

                // Draw selection indicator
                if i as i32 == self.selected {
                    let indicator_x = start_x + name.len() + 1;
                    screen.put_str(indicator_x, y, "<");
                    screen.set_fg_color(indicator_x, y, Color::YELLOW);
                }
            }
        }
    }
}

impl Default for Menu {
    fn default() -> Self {
        Self::new()
    }
}
